pandas/_libs/window.pyx

```diff
@@ -1414,18 +1438,42 @@ def roll_quantile(ndarray[float64_t, cast=True] input, int64_t win,
                     skiplist.insert(val)
 
         if nobs >= minp:
-            idx = int(quantile * <double>(nobs - 1))
-
-            # Single value in skip list
             if nobs == 1:
+                # Single value in skip list
                 output[i] = skiplist.get(0)
-
-            # Interpolated quantile
             else:
-                vlow = skiplist.get(idx)
-                vhigh = skiplist.get(idx + 1)
-                output[i] = ((vlow + (vhigh - vlow) *
-                             (quantile * (nobs - 1) - idx)))
+                idx_with_fraction = quantile * (nobs - 1)
+                idx = <int> idx_with_fraction
+
+                if idx_with_fraction == idx:
+                    # no need to interpolate
+                    output[i] = skiplist.get(idx)
+                    continue
+
+                if interpolation_type == LINEAR:
+                    vlow = skiplist.get(idx)
+                    vhigh = skiplist.get(idx + 1)
+                    output[i] = ((vlow + (vhigh - vlow) *
+                                  (idx_with_fraction - idx)))
+                elif interpolation_type == LOWER:
+                    output[i] = skiplist.get(idx)
+                elif interpolation_type == HIGHER:
+                    output[i] = skiplist.get(idx + 1)
+                elif interpolation_type == NEAREST:
+                    # the same behaviour as round()
+                    if idx_with_fraction - idx == 0.5:
+                        if idx % 2 == 0:
+                            output[i] = skiplist.get(idx)
+                        else:
+                            output[i] = skiplist.get(idx + 1)
+                    elif idx_with_fraction - idx < 0.5:
+                        output[i] = skiplist.get(idx)
+                    else:
+                        output[i] = skiplist.get(idx + 1)
+                elif interpolation_type == MIDPOINT:
+                    vlow = skiplist.get(idx)
```

Comment:
to make this work w/o the gil,  you need to call ``skiplist_get(skiplist, idx)`` (there are examples in the file). ``skiplist.get`` invokes a python function call and is not allowed